utils/rowgroup/rowstorage.cpp

```diff
@@ -0,0 +1,2139 @@
+#include <resourcemanager.h>
+#include <fcntl.h>
+#include "rowstorage.h"
+#include "robin_hood.h"
+
+namespace
+{
+
+bool writeData(int fd, const char* buf, size_t sz)
+{
+  if (sz == 0)
+    return true;
+
+  auto to_write = sz;
+  while (to_write > 0)
+  {
+    auto r = write(fd, buf + sz - to_write, to_write);
+    if (UNLIKELY(r < 0))
+    {
+      if (errno == EAGAIN)
+        continue;
+
+      return false;
+    }
+    assert(size_t(r) <= to_write);
+    to_write -= r;
+  }
+
+  return true;
+}
+
+bool readData(int fd, char* buf, size_t sz)
+{
+  if (sz == 0)
+    return true;
+
+  auto to_read = sz;
+  while (to_read > 0)
+  {
+    auto r = read(fd, buf + sz - to_read, to_read);
+    if (UNLIKELY(r < 0))
+    {
+      if (errno == EAGAIN)
+        continue;
+
+      return false;
+    }
+
+    assert(size_t(r) <= to_read);
+    to_read -= r;
+  }
+
+  return true;
+}
+
+inline uint64_t hashData(const void* ptr, uint32_t len, uint64_t x = 0ULL)
+{
+  static constexpr uint64_t m = 0xc6a4a7935bd1e995ULL;
+  static constexpr uint64_t seed = 0xe17a1465ULL;
+  static constexpr unsigned int r = 47;
+
+  auto const* const data64 = static_cast<uint64_t const*>(ptr);
+  uint64_t h = seed ^ (len * m);
+
+  std::size_t const n_blocks = len / 8;
+  if (x)
+  {
+    x *= m;
+    x ^= x >> r;
+    x *= m;
+    h ^= x;
+    h *= m;
+  }
+  for (std::size_t i = 0; i < n_blocks; ++i) {
+    uint64_t k;
+    memcpy(&k, data64 + i, sizeof(k));
+
+    k *= m;
+    k ^= k >> r;
+    k *= m;
+
+    h ^= k;
+    h *= m;
+  }
+
+  auto const* const data8 = reinterpret_cast<uint8_t const*>(data64 + n_blocks);
+  switch (len & 7U) {
+  case 7:
+    h ^= static_cast<uint64_t>(data8[6]) << 48U;
+    // FALLTHROUGH
+  case 6:
+    h ^= static_cast<uint64_t>(data8[5]) << 40U;
+    // FALLTHROUGH
+  case 5:
+    h ^= static_cast<uint64_t>(data8[4]) << 32U;
+    // FALLTHROUGH
+  case 4:
+    h ^= static_cast<uint64_t>(data8[3]) << 24U;
+    // FALLTHROUGH
+  case 3:
+    h ^= static_cast<uint64_t>(data8[2]) << 16U;
+    // FALLTHROUGH
+  case 2:
+    h ^= static_cast<uint64_t>(data8[1]) << 8U;
+    // FALLTHROUGH
+  case 1:
+    h ^= static_cast<uint64_t>(data8[0]);
+    h *= m;
+    // FALLTHROUGH
+  default:
+    break;
+  }
+
+  h ^= h >> r;
+  h *= m;
+  h ^= h >> r;
+
+  return h;
+}
+
+inline uint64_t hashRow(const rowgroup::Row& r, std::size_t lastCol)
+{
+  uint64_t ret = 0;
+  if (lastCol >= r.getColumnCount())
+    return 0;
+
+  datatypes::MariaDBHasher h;
+  for (uint32_t i = 0; i <= lastCol; ++i)
+  {
+    switch (r.getColType(i))
+    {
+    case execplan::CalpontSystemCatalog::CHAR:
+    case execplan::CalpontSystemCatalog::VARCHAR:
+    case execplan::CalpontSystemCatalog::BLOB:
+    case execplan::CalpontSystemCatalog::TEXT:
+      h.add(r.getCharset(i), r.getConstString(i));
+      break;
+
+    default:
+      ret = hashData(r.getData() + r.getOffset(i), r.getColumnWidth(i), ret);
+      break;
+    }
+  }
+
+  if (UNLIKELY(h.wasUsed()))
```

Comment:
I don't think this condition is unlikely for a whole Row.

Reply:
Yes, you're right.